File: head_tracking_audio/python_host/dsp/hrtf_selector.py
import numpy as np


def _nearest_index(values, target):
	values = np.asarray(values).reshape(-1)
	if values.size == 0:
		raise ValueError("Empty angle array")
	return int(np.argmin(np.abs(values - target)))


def _get_hrir_at(dataset, az_idx, el_idx):
	hrir_l = dataset.hrir_l
	hrir_r = dataset.hrir_r

	if hrir_l.ndim != 3:
		raise ValueError("Expected 3D HRIR arrays")

	if hrir_l.shape[0] == dataset.elevations.size and hrir_l.shape[1] == dataset.azimuths.size:
		left = hrir_l[el_idx, az_idx, :]
		right = hrir_r[el_idx, az_idx, :]
	elif hrir_l.shape[0] == dataset.azimuths.size and hrir_l.shape[1] == dataset.elevations.size:
		left = hrir_l[az_idx, el_idx, :]
		right = hrir_r[az_idx, el_idx, :]
	else:
		raise ValueError("HRIR shape does not match azimuth/elevation vectors")

	return left, right


def select_hrir(dataset, yaw_deg, pitch_deg=0.0):
	az_idx = _nearest_index(dataset.azimuths, yaw_deg)
	el_idx = _nearest_index(dataset.elevations, pitch_deg)

	left, right = _get_hrir_at(dataset, az_idx, el_idx)
	used = (float(dataset.azimuths[az_idx]), float(dataset.elevations[el_idx]))
	return left, right, used


def select_hrir_interpolated(dataset, yaw_deg, pitch_deg=0.0):
	az = np.asarray(dataset.azimuths).reshape(-1)
	order = np.argsort(az)
	az_sorted = az[order]

	if az_sorted.size == 0:
		raise ValueError("Empty azimuth array")

	if yaw_deg <= az_sorted[0]:
		idx0 = idx1 = order[0]
		alpha = 0.0
	elif yaw_deg >= az_sorted[-1]:
		idx0 = idx1 = order[-1]
		alpha = 0.0
	else:
		insert = int(np.searchsorted(az_sorted, yaw_deg))
		idx0 = order[insert - 1]
		idx1 = order[insert]
		az0 = float(az_sorted[insert - 1])
		az1 = float(az_sorted[insert])
		alpha = 0.0 if az1 == az0 else (yaw_deg - az0) / (az1 - az0)

	el_idx = _nearest_index(dataset.elevations, pitch_deg)
	left0, right0 = _get_hrir_at(dataset, idx0, el_idx)
	left1, right1 = _get_hrir_at(dataset, idx1, el_idx)

	left = (1.0 - alpha) * left0 + alpha * left1
	right = (1.0 - alpha) * right0 + alpha * right1

	used = (float(dataset.azimuths[idx0]), float(dataset.azimuths[idx1]), float(dataset.elevations[el_idx]))
	return left, right, used, float(alpha)

File: head_tracking_audio/tools/matlab/hrtf_selector.m
function [left,right,used] = select_hrir(dataset, yaw_deg, pitch_deg)
% SELECT_HRIR nearest selection (no interpolation)
az = double(dataset.azimuths(:));
el = double(dataset.elevations(:));
[~, az_idx] = min(abs(az - yaw_deg));
[~, el_idx] = min(abs(el - pitch_deg));
hrir_l = dataset.hrir_l;
hrir_r = dataset.hrir_r;
% assume hrir dims elev x az x samples OR az x elev x samples -> detect
if size(hrir_l,1) == numel(el) && size(hrir_l,2) == numel(az)
    left = squeeze(hrir_l(el_idx, az_idx, :))';
    right = squeeze(hrir_r(el_idx, az_idx, :))';
elseif size(hrir_l,1) == numel(az) && size(hrir_l,2) == numel(el)
    left = squeeze(hrir_l(az_idx, el_idx, :))';
    right = squeeze(hrir_r(az_idx, el_idx, :))';
else
    error('HRIR shape does not match azimuth/elevation vectors');
end
used = [az(az_idx), el(el_idx)];
end

function [left,right,used0,used1,alpha] = select_hrir_interpolated(dataset, yaw_deg, pitch_deg)
az = double(dataset.azimuths(:));
el = double(dataset.elevations(:));
[~, el_idx] = min(abs(el - pitch_deg));
% sort azimuths
[az_sorted, order] = sort(az);
if yaw_deg <= az_sorted(1)
    idx0 = order(1); idx1 = order(1); alpha = 0;
elseif yaw_deg >= az_sorted(end)
    idx0 = order(end); idx1 = order(end); alpha = 0;
else
    insert = find(az_sorted >= yaw_deg,1);
    idx0 = order(insert-1);
    idx1 = order(insert);
    az0 = az_sorted(insert-1); az1 = az_sorted(insert);
    if az1 == az0
        alpha = 0;
    else
        alpha = (yaw_deg - az0) / (az1 - az0);
    end
end
% get HRIRs
[l0,r0] = get_hrir(dataset, idx0, el_idx);
[l1,r1] = get_hrir(dataset, idx1, el_idx);
left = (1-alpha)*l0 + alpha*l1;
right = (1-alpha)*r0 + alpha*r1;
used0 = az(idx0); used1 = az(idx1);
end

function [l,r] = get_hrir(dataset, az_idx, el_idx)
    hrir_l = dataset.hrir_l; hrir_r = dataset.hrir_r;
    if size(hrir_l,1) == numel(dataset.elevations) && size(hrir_l,2) == numel(dataset.azimuths)
        l = squeeze(hrir_l(el_idx, az_idx, :))';
        r = squeeze(hrir_r(el_idx, az_idx, :))';
    elseif size(hrir_l,1) == numel(dataset.azimuths) && size(hrir_l,2) == numel(dataset.elevations)
        l = squeeze(hrir_l(az_idx, el_idx, :))';
        r = squeeze(hrir_r(az_idx, el_idx, :))';
    else
        error('Unexpected HRIR layout');
    end
end
